File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-6.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    goodNodes(root: TreeNode | null): number {
        const dfs = (root: TreeNode | null, maxSeen: number): number => {
            if (!root) return 0
            const count = root.val >= maxSeen ? 1 : 0
            maxSeen = Math.max(maxSeen, root.val)

            return count + dfs(root.left, maxSeen) + dfs(root.right, maxSeen)
        }

        return dfs(root, root.val)
    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-6.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.empty() || inorder.empty()) return nullptr;

        TreeNode* root = new TreeNode(preorder[0]);
        int inorder_root_index = find(inorder.begin(), inorder.end(), root->val) - inorder.begin();
        vector<int> preorder_left(preorder.begin() + 1, preorder.begin() + 1 + inorder_root_index);
        vector<int> preorder_right(preorder.begin() + inorder_root_index + 1, preorder.end());
        vector<int> inorder_left(inorder.begin(), inorder.begin() + inorder_root_index);
        vector<int> inorder_right(inorder.begin() + inorder_root_index + 1, inorder.end());

        root->left = buildTree(preorder_left, inorder_left);
        root->right = buildTree(preorder_right, inorder_right);

        return root;
    }
};
